File: quick_sort_step.java
import java.io.*;
import java.util.*;

public class quick_sort_step {

    static class Element {
        int number;
        String word;

        Element(int number, String word) {
            this.number = number;
            this.word = word;
        }

        public String toString() {
            return number + "/" + word;
        }
    }

    public static void main(String[] args) {
        if (args.length != 3) {
            System.out.println("Usage: java QuickSortStep dataset_sample_1000.csv startRow endRow");
            return;
        }

        String inputFile = args[0];
        int startRow = Integer.parseInt(args[1]);
        int endRow = Integer.parseInt(args[2]);

        try {
            List<Element> data = readSubset(inputFile, startRow, endRow);
            List<String> stepLog = new ArrayList<>();
            stepLog.add(formatList(data)); // Initial state
            quickSort(data, 0, data.size() - 1, stepLog);
            writeSteps("quick_sort_step_" + startRow + "_" + endRow + ".txt", stepLog);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    static List<Element> readSubset(String filename, int startRow, int endRow) throws IOException {
        List<Element> list = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            int row = 1;
            while ((line = reader.readLine()) != null) {
                if (row >= startRow && row <= endRow) {
                    String[] parts = line.split(",", 2);
                    list.add(new Element(Integer.parseInt(parts[0]), parts[1]));
                }
                row++;
            }
        }
        return list;
    }

    static void writeSteps(String filename, List<String> steps) throws IOException {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            for (String step : steps) {
                writer.write(step);
                writer.newLine();
            }
        }
    }

    static void quickSort(List<Element> arr, int low, int high, List<String> log) {
        if (low < high) {
            int pi = partition(arr, low, high);
            log.add("pi=" + pi + " " + formatList(arr));
            quickSort(arr, low, pi - 1, log);
            quickSort(arr, pi + 1, high, log);
        }
    }

    static int partition(List<Element> arr, int low, int high) {
        int pivot = arr.get(high).number;
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr.get(j).number < pivot) {
                i++;
                Collections.swap(arr, i, j);
            }
        }
        Collections.swap(arr, i + 1, high);
        return i + 1;
    }

    static String formatList(List<Element> arr) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < arr.size(); i++) {
            sb.append(arr.get(i));
            if (i < arr.size() - 1) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }
}

File: quick_sort_step.py
import csv
import sys

#quick_sort modify sikit so boleh user friendly terima input lain
def read_csv_subset(filename, start_row, end_row):
    with open(filename, newline='') as csvfile:
        reader = csv.reader(csvfile)
        data = [row for i, row in enumerate(reader) if start_row <= i+1 <= end_row]
    return [(int(row[0]), row[1]) for row in data]

def write_steps(filename, steps):
    with open(filename, 'w') as f:
        for step in steps:
            f.write(step + '\n')

def quick_sort(arr, low, high, steps):
    if low < high:
        pi = partition(arr, low, high, steps)
        steps.append(f"pi={pi} {format_list(arr)}")
        quick_sort(arr, low, pi - 1, steps)
        quick_sort(arr, pi + 1, high, steps)

def partition(arr, low, high, steps):
    pivot = arr[high][0]
    i = low - 1
    for j in range(low, high):
        if arr[j][0] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i+1], arr[high] = arr[high], arr[i+1]
    return i + 1

def format_list(arr):
    return "[" + ", ".join(f"{x[0]}/{x[1]}" for x in arr) + "]"

def main():
    if len(sys.argv) != 4:
        print("Usage: python quick_sort_step.py dataset_sample_1000.csv start_row end_row")
        return

    filename, start_row, end_row = sys.argv[1], int(sys.argv[2]), int(sys.argv[3])
    data = read_csv_subset(filename, start_row, end_row)
    steps = [format_list(data)]
    quick_sort(data, 0, len(data) - 1, steps)
    output_file = f"quick_sort_step_{start_row}_{end_row}.txt"
    write_steps(output_file, steps)

if __name__ == "__main__":
    main()